src/persist/src/indexed/mod.rs

```diff
@@ -126,15 +126,31 @@ impl<K: Data, V: Data, U: Buffer, L: Blob> Indexed<K, V, U, L> {
             blob,
             futures,
             traces,
-            listeners: HashMap::new(),
+            listeners: BTreeMap::new(),
         };
+
+        indexed.check_invariants()?;
         Ok(indexed)
     }
 
+    /// Sanity check invariants at runtime.
+    fn check_invariants(&self) -> Result<(), Error> {
+        if cfg!(any(test, debug)) {
+            let stored_meta = self.blob.get_meta()?.unwrap_or_default();
+            let local_meta = self.serialize_meta();
+
+            assert_eq!(stored_meta, local_meta);
+            local_meta.validate()?;
+        }
+
+        Ok(())
+    }
+
     /// Releases exclusive-writer locks and causes all future commands to error.
     ///
     /// This method is idempotent.
     pub fn close(&mut self) -> Result<(), Error> {
+        self.check_invariants()?;
```

Comment:
why not a debug_assert on all these? this one might be fine getting called in prod, but stuff like check_invariants and sealed_frontier really aren't

Reply:
I understand that. I thought the `cfg!(any(test, debug))` in `check_invariants` would guard against this being called on an optimized non-test execution, but make it so it still checks invariants even if you run tests on an optimized build, which I felt was useful enough to warrant doing it this way.

However, just looking around now I can't find any reference that says `debug` is the right attribute name. It seems like it might be `debug_assertions` instead?

Reply:
Ah, gotcha! So first, I'd prefer this be at the callsite, so a reader doesn't have to follow the code down to see that. Second, I prefer to avoid having to reason about different sets of possible return points in test vs release (even if that reasoning is very straightforward like here). So my new counter proposal:

```
if cfg!(any(test, debug_assertions)) {
    assert_eq!(self.check_invariants(), Ok(()));
}
```

possibly with a `, "{:?}", self.serialize_meta()` if we think that's useful.

And then turn the assert_eq! inside the impl of check_invariants into a `if stored != local { return Err(...) }` type thing

> However, just looking around now I can't find any reference that says debug is the right attribute name. It seems like it might be debug_assertions instead?

You're right, it's definitely debug_assertions. I just ran a little test and it seems that it'll accept anything there, so I think "debug" really isn't anything. I think I introduced that one, so my bad :)